File: src/main/java/codereview/simpleorder/item/application/ItemService.java
package codereview.simpleorder.item.application;

import codereview.simpleorder.item.domain.Item;
import codereview.simpleorder.item.dto.web.CreateItemRequest;
import codereview.simpleorder.item.dto.web.ItemResponse;
import codereview.simpleorder.item.repository.ItemRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;
import java.util.Set;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class ItemService {

    private final ItemRepository itemRepository;

    @Transactional
    public Long registerItem(CreateItemRequest request) {

        Item item = createItem(request);
        Item savedItem = itemRepository.save(item);

        return savedItem.getId();
    }

    public List<ItemResponse> findItems(List<Long> itemIds) {

        if (itemIds == null || itemIds.isEmpty()) {

            return itemRepository.findAll().stream()
                    .map(ItemResponse::new)
                    .toList();
        }

        return itemRepository.findByIdIn(itemIds).stream()
                .map(ItemResponse::new)
                .toList();
    }

    @Transactional
    public void decreaseItems(Map<Long, Integer> itemIdToDecreaseQuantity) {

        Set<Long> itemIds = itemIdToDecreaseQuantity.keySet();
        List<Item> findItems = itemRepository.findByIdIn(itemIds); // problem !
        validateAndDecreaseItemQuantity(itemIdToDecreaseQuantity, findItems);
    }

    private static Item createItem(CreateItemRequest request) {

        return new Item(request.getName(), request.getSize(), request.getPrice(), request.getAvailableQuantity());
    }

    private static void validateAndDecreaseItemQuantity(Map<Long, Integer> itemIdToDecreaseQuantity, List<Item> findItems) {

        for (Item item : findItems) {
            int decreaseQuantity = itemIdToDecreaseQuantity.get(item.getId());
            item.decreaseQuantity(decreaseQuantity);
        }
    }
}

File: legacy/src/main/kotlin/msa/with/ddd/item/application/ItemService.kt
package msa.with.ddd.item.application

import msa.with.ddd.item.domain.Item
import msa.with.ddd.item.dto.web.CreateItemRequest
import msa.with.ddd.item.dto.web.ItemResponse
import msa.with.ddd.item.repository.ItemRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class ItemService(private val itemRepository: ItemRepository) {

    @Transactional
    fun registerItem(request: CreateItemRequest): Long {
        val item = createItem(request)
        val savedItem = itemRepository.save(item)
        return savedItem.id!!
    }

    fun findItems(itemIds: List<Long>?): List<ItemResponse> {

        return if (itemIds.isNullOrEmpty()) {
            itemRepository.findAll()
                .map { item -> ItemResponse(item) }
                .toList()
        } else {
            itemRepository.findByIdIn(itemIds)
                .map { item -> ItemResponse(item) }
                .toList()
        }
    }

    @Transactional
    fun decreaseItems(itemIdToDecreaseQuantity: Map<Long, Int>) {
        val itemIds = itemIdToDecreaseQuantity.keys
        val findItems = itemRepository.findByIdIn(itemIds) // problem !
        validateAndDecreaseItemQuantity(itemIdToDecreaseQuantity, findItems)
    }

    private fun createItem(request: CreateItemRequest) =
        Item(request.name, request.size, request.price, request.availableQuantity)

    private fun validateAndDecreaseItemQuantity(itemIdToDecreaseQuantity: Map<Long, Int>, findItems: List<Item>) {

        for (item in findItems) {
            val decreaseQuantity = itemIdToDecreaseQuantity[item.id]!!
            item.decreaseQuantity(decreaseQuantity)
        }
    }
}